internal/p2p/channel_test.go

```diff
@@ -0,0 +1,183 @@
+				require.NoError(t, ch.SendError(ctx, PeerError{NodeID: "kip", Err: errors.New("merlin")}))
+
+				res, ok := <-ins.Error
+				require.True(t, ok)
+				require.EqualValues(t, "kip", res.NodeID)
+				require.EqualValues(t, "merlin", res.Err.Error())
+			},
+		},
+		{
+			Name: "SendWithCanceledContext",
+			Case: func(ctx context.Context, t *testing.T) {
+				_, ch := testChannel(0)
+				cctx, ccancel := context.WithCancel(ctx)
+				ccancel()
+				require.Error(t, ch.Send(cctx, Envelope{From: "kip", To: "merlin"}))
+			},
+		},
+		{
+			Name: "SendErrorWithCanceledContext",
+			Case: func(ctx context.Context, t *testing.T) {
+				_, ch := testChannel(0)
+				cctx, ccancel := context.WithCancel(ctx)
+				ccancel()
+
+				require.Error(t, ch.SendError(cctx, PeerError{NodeID: "kip", Err: errors.New("merlin")}))
+			},
+		},
+		{
+			Name: "ReceiveEmptyIteratorBlocks",
+			Case: func(ctx context.Context, t *testing.T) {
+				_, ch := testChannel(1)
+				iter := ch.Receive(ctx)
+				require.NotNil(t, iter)
+				out := make(chan bool)
+				go func() {
+					defer close(out)
+					select {
+					case <-ctx.Done():
+					case out <- iter.Next(ctx):
+					}
+				}()
+				select {
+				case <-time.After(10 * time.Millisecond):
+				case <-out:
+					require.Fail(t, "iterator should not advance")
+				}
+				require.Nil(t, iter.Envelope())
+			},
+		},
+		{
+			Name: "ReceiveWithData",
+			Case: func(ctx context.Context, t *testing.T) {
+				ins, ch := testChannel(1)
+				ins.In <- Envelope{From: "kip", To: "merlin"}
+				iter := ch.Receive(ctx)
+				require.NotNil(t, iter)
+				require.True(t, iter.Next(ctx))
+
+				res := iter.Envelope()
+				require.EqualValues(t, "kip", res.From)
+				require.EqualValues(t, "merlin", res.To)
+			},
+		},
+		{
+			Name: "ReceiveWithCanceledContext",
+			Case: func(ctx context.Context, t *testing.T) {
+				_, ch := testChannel(0)
+				cctx, ccancel := context.WithCancel(ctx)
+				ccancel()
+
+				iter := ch.Receive(cctx)
+				require.NotNil(t, iter)
+				require.False(t, iter.Next(cctx))
+				require.Nil(t, iter.Envelope())
+			},
+		},
+		{
+			Name: "IteratorWithCanceledContext",
+			Case: func(ctx context.Context, t *testing.T) {
+				_, ch := testChannel(0)
+
+				iter := ch.Receive(ctx)
+				require.NotNil(t, iter)
+
+				cctx, ccancel := context.WithCancel(ctx)
+				ccancel()
+				require.False(t, iter.Next(cctx))
+				require.Nil(t, iter.Envelope())
+			},
+		},
+		{
+			Name: "IteratorCanceledAfterFirstUseBecomesNil",
+			Case: func(ctx context.Context, t *testing.T) {
+				ins, ch := testChannel(1)
+
+				ins.In <- Envelope{From: "kip", To: "merlin"}
+				iter := ch.Receive(ctx)
+				require.NotNil(t, iter)
+
+				require.True(t, iter.Next(ctx))
+
+				res := iter.Envelope()
+				require.EqualValues(t, "kip", res.From)
+				require.EqualValues(t, "merlin", res.To)
+
+				cctx, ccancel := context.WithCancel(ctx)
+				ccancel()
+
+				require.False(t, iter.Next(cctx))
+				require.Nil(t, iter.Envelope())
```

Comment:
Really appreciate these tests. I like the semantics of having `Next` nil out the value when the ctx is canceled.